### ISSUE TITLE
Bug: the "args": "after-used" is not respected in some cases

### ISSUE THREAD

**Issue by rusproject** [eslint/eslint#17639]
### Environment

Node version: 18.17.1
npm version: 10.1.0
Local ESLint version: 8.45.0
Global ESLint version: 8.45.0
Operating System: Windows 10 x64

### What parser are you using?

Default (Espree)

### What did you do?

<details>
<summary>Configuration</summary>

```json
{
  "parserOptions": {
    "ecmaVersion": "latest"
  },
  "env": {
    "browser": true
  },
  "rules": {
    "no-unused-vars": ["warn", { "args": "after-used" }]
  }
}
```
</details>

```js
// OK: eslint respects "args": "after-used"

function someFunc(arg1, arg2) {
  return arg2;
}

console.log(someFunc());


/******/


// FAIL: eslint ignores "args": "after-used" and warns that 'key' is defined but never used

const array = [[1, 'a'], [2, 'b']];

array.forEach(([key, value]) => {
  console.log(value);
});
```

### What did you expect to happen?

`"no-unused-vars": ["warn", { "args": "after-used" }]` is respected and eslint doesn't reports any warnings of the `no-unused-vars` type for this code.

### What actually happened?

ESlint ignores `"args": "after-used"` and warns that 'key' is defined but never used (although it does respect this option for the first function and everything is fine there, as you can see on the screenshot).

![image](https://github.com/eslint/eslint/assets/62891105/e3e4025e-a9e7-42b2-84ce-43f25646f105)

### Link to Minimal Reproducible Example

https://eslint.org/play/#eyJ0ZXh0IjoiLy8gT0s6IEVTTGludCByZXNwZWN0cyBcImFyZ3NcIjogXCJhZnRlci11c2VkXCJcbmZ1bmN0aW9uIHNvbWVGdW5jKGFyZzEsIGFyZzIpIHtcbiAgcmV0dXJuIGFyZzI7XG59XG5zb21lRnVuYygpO1xuXG4vLyBGQUlMOiBFU2xpbnQgaWdub3JlcyBcImFyZ3NcIjogXCJhZnRlci11c2VkXCJcbmNvbnN0IGFycmF5ID0gW1sxLCAnYSddLCBbMiwgJ2InXV07XG5cbmFycmF5LmZvckVhY2goKFtrZXksIHZhbHVlXSkgPT4ge1xuICB2YWx1ZTtcbn0pO1xuIiwib3B0aW9ucyI6eyJlbnYiOnsiZXM2Ijp0cnVlfSwicGFyc2VyT3B0aW9ucyI6eyJlY21hRmVhdHVyZXMiOnt9LCJlY21hVmVyc2lvbiI6ImxhdGVzdCIsInNvdXJjZVR5cGUiOiJzY3JpcHQifSwicnVsZXMiOnsiY29uc3RydWN0b3Itc3VwZXIiOlsiZXJyb3IiXSwiZm9yLWRpcmVjdGlvbiI6WyJlcnJvciJdLCJnZXR0ZXItcmV0dXJuIjpbImVycm9yIl0sIm5vLWFzeW5jLXByb21pc2UtZXhlY3V0b3IiOlsiZXJyb3IiXSwibm8tY2FzZS1kZWNsYXJhdGlvbnMiOlsiZXJyb3IiXSwibm8tY2xhc3MtYXNzaWduIjpbImVycm9yIl0sIm5vLWNvbXBhcmUtbmVnLXplcm8iOlsiZXJyb3IiXSwibm8tY29uZC1hc3NpZ24iOlsiZXJyb3IiXSwibm8tY29uc3QtYXNzaWduIjpbImVycm9yIl0sIm5vLWNvbnN0YW50LWNvbmRpdGlvbiI6WyJlcnJvciJdLCJuby1jb250cm9sLXJlZ2V4IjpbImVycm9yIl0sIm5vLWRlYnVnZ2VyIjpbImVycm9yIl0sIm5vLWRlbGV0ZS12YXIiOlsiZXJyb3IiXSwibm8tZHVwZS1hcmdzIjpbImVycm9yIl0sIm5vLWR1cGUtY2xhc3MtbWVtYmVycyI6WyJlcnJvciJdLCJuby1kdXBlLWVsc2UtaWYiOlsiZXJyb3IiXSwibm8tZHVwZS1rZXlzIjpbImVycm9yIl0sIm5vLWR1cGxpY2F0ZS1jYXNlIjpbImVycm9yIl0sIm5vLWVtcHR5IjpbImVycm9yIl0sIm5vLWVtcHR5LWNoYXJhY3Rlci1jbGFzcyI6WyJlcnJvciJdLCJuby1lbXB0eS1wYXR0ZXJuIjpbImVycm9yIl0sIm5vLWV4LWFzc2lnbiI6WyJlcnJvciJdLCJuby1leHRyYS1ib29sZWFuLWNhc3QiOlsiZXJyb3IiXSwibm8tZXh0cmEtc2VtaSI6WyJlcnJvciJdLCJuby1mYWxsdGhyb3VnaCI6WyJlcnJvciJdLCJuby1mdW5jLWFzc2lnbiI6WyJlcnJvciJdLCJuby1nbG9iYWwtYXNzaWduIjpbImVycm9yIl0sIm5vLWltcG9ydC1hc3NpZ24iOlsiZXJyb3IiXSwibm8taW5uZXItZGVjbGFyYXRpb25zIjpbImVycm9yIl0sIm5vLWludmFsaWQtcmVnZXhwIjpbImVycm9yIl0sIm5vLWlycmVndWxhci13aGl0ZXNwYWNlIjpbImVycm9yIl0sIm5vLWxvc3Mtb2YtcHJlY2lzaW9uIjpbImVycm9yIl0sIm5vLW1pc2xlYWRpbmctY2hhcmFjdGVyLWNsYXNzIjpbImVycm9yIl0sIm5vLW1peGVkLXNwYWNlcy1hbmQtdGFicyI6WyJlcnJvciJdLCJuby1uZXctc3ltYm9sIjpbImVycm9yIl0sIm5vLW5vbm9jdGFsLWRlY2ltYWwtZXNjYXBlIjpbImVycm9yIl0sIm5vLW9iai1jYWxscyI6WyJlcnJvciJdLCJuby1vY3RhbCI6WyJlcnJvciJdLCJuby1wcm90b3R5cGUtYnVpbHRpbnMiOlsiZXJyb3IiXSwibm8tcmVkZWNsYXJlIjpbImVycm9yIl0sIm5vLXJlZ2V4LXNwYWNlcyI6WyJlcnJvciJdLCJuby1zZWxmLWFzc2lnbiI6WyJlcnJvciJdLCJuby1zZXR0ZXItcmV0dXJuIjpbImVycm9yIl0sIm5vLXNoYWRvdy1yZXN0cmljdGVkLW5hbWVzIjpbImVycm9yIl0sIm5vLXNwYXJzZS1hcnJheXMiOlsiZXJyb3IiXSwibm8tdGhpcy1iZWZvcmUtc3VwZXIiOlsiZXJyb3IiXSwibm8tdW5kZWYiOlsiZXJyb3IiXSwibm8tdW5leHBlY3RlZC1tdWx0aWxpbmUiOlsiZXJyb3IiXSwibm8tdW5yZWFjaGFibGUiOlsiZXJyb3IiXSwibm8tdW5zYWZlLWZpbmFsbHkiOlsiZXJyb3IiXSwibm8tdW5zYWZlLW5lZ2F0aW9uIjpbImVycm9yIl0sIm5vLXVuc2FmZS1vcHRpb25hbC1jaGFpbmluZyI6WyJlcnJvciJdLCJuby11bnVzZWQtbGFiZWxzIjpbImVycm9yIl0sIm5vLXVudXNlZC12YXJzIjpbImVycm9yIix7ImFyZ3MiOiJhZnRlci11c2VkIn1dLCJuby11c2VsZXNzLWJhY2tyZWZlcmVuY2UiOlsiZXJyb3IiXSwibm8tdXNlbGVzcy1jYXRjaCI6WyJlcnJvciJdLCJuby11c2VsZXNzLWVzY2FwZSI6WyJlcnJvciJdLCJuby13aXRoIjpbImVycm9yIl0sInJlcXVpcmUteWllbGQiOlsiZXJyb3IiXSwidXNlLWlzbmFuIjpbImVycm9yIl0sInZhbGlkLXR5cGVvZiI6WyJlcnJvciJdfX19

### Participation

- [ ] I am willing to submit a pull request for this issue.

**Comment by rusproject** [comment]
My bad. Of course, `key` and `value` in the `forEach(([key, value])...` are not arguments, since it is array destructuring and they are considered as just variables declaration. So the ESLint behavior is expected I suppose.

However, since we can't just omit the `key` in such cases, it'd be nice if  `"args": "after-used"` treated such declarations in the same way it does for function arguments: if the first one is obligatory and the second one is used, then it reports no error with that option. This way, we could avoid writing code like this:
```js
array.forEach(([, value]) => {
  console.log(value);
});
```

**Comment by Rec0iL99** [comment]
Hi @rusproject, thanks for the issue. This looks like a duplicate of #10543. 

Your example in https://github.com/eslint/eslint/issues/17639#issuecomment-1756048429 is the right way to skip the `key` variable. You can find a clear explanation for this at https://github.com/eslint/eslint/issues/10543#issuecomment-401427801.

If you don't like skipping a variable in array destructuring, then you can use the `destructuredArrayIgnorePattern` option in the rule to ignore the variable using a regexp pattern.

```js
/* eslint "no-unused-vars": ["error", { "args": "after-used", "destructuredArrayIgnorePattern": "^_" }] */

const array = [[1, 'a'], [2, 'b']];

array.forEach(([_key, value]) => {
  value;
});
``` 

[playground](https://eslint.org/play/#eyJ0ZXh0IjoiLyogZXNsaW50IFwibm8tdW51c2VkLXZhcnNcIjogW1wiZXJyb3JcIiwgeyBcImFyZ3NcIjogXCJhZnRlci11c2VkXCIsIFwiZGVzdHJ1Y3R1cmVkQXJyYXlJZ25vcmVQYXR0ZXJuXCI6IFwiXl9cIiB9XSAqL1xuXG5jb25zdCBhcnJheSA9IFtbMSwgJ2EnXSwgWzIsICdiJ11dO1xuXG5hcnJheS5mb3JFYWNoKChbX2tleSwgdmFsdWVdKSA9PiB7XG4gIHZhbHVlO1xufSk7XG4iLCJvcHRpb25zIjp7ImVudiI6eyJlczYiOnRydWV9LCJydWxlcyI6e30sInBhcnNlck9wdGlvbnMiOnsiZWNtYUZlYXR1cmVzIjp7fSwiZWNtYVZlcnNpb24iOiJsYXRlc3QiLCJzb3VyY2VUeXBlIjoic2NyaXB0In19fQ==)

**Comment by rusproject** [comment]
@Rec0iL99 Got it! Thanks for your response and for the example with `destructuredArrayIgnorePattern`. Good to know that this is not something missed out, but rather an intentional design.
